keps/sig-apps/2804-consolidate-workload-controllers-status/README.md

```diff
@@ -0,0 +1,370 @@
+- DaemonSet and StatefulSet controllers can be marked `Failed` even though rollout is in progress
+- The states could be misrepresented, for example a DaemonSet or StatefulSet can be marked `Progressing` when actually it is complete
+
+The mitigation currently is that these features will be in Alpha stage behind `ExtendedWorkloadConditions` featuregate for people to try out and give feedback. In Beta phase when
+these are enabled by default, people will only see issues or bugs when `progressDeadlineSeconds` is set to something greater than 0 and we choose default values for that field.
+Since people would have tried this feature in Alpha, we would have had time to fix issues.
+
+
+## Design Details
+
+### Test Plan
+Unit and E2E tests will be added to cover the
+API validation, behavioral change of controllers with feature gate enabled and disabled.
+
+- Validating all possible states of old and new conditions. Checking that the changes in underlying Pod statuses correspond to the conditions.
+- Testing `progressDeadlineSeconds` and feature gates.
+
+
+### Graduation Criteria
+
+#### Alpha
+- Complete feature behind featuregates
+- Have proper unit, integration and e2e tests 
+
+#### Alpha -> Beta Graduation
+- Gather feedback from end users
+- Tests are in Testgrid and linked in KEP
+- all new features in the following controllers should be implemented: ReplicaSet & ReplicationController, StatefulSet, DaemonSet and Job. To fully support `failed` state of a progressing condition, `progressDeadlineSeconds` KEP should be also fully implemented.
+
+#### Beta -> GA Graduation
+- 2 examples of end users using this field
+
+### Upgrade / Downgrade Strategy
+
+- Upgrades
+ When upgrading from a release without this feature, to a release with `ExtendedWorkloadConditions` feature,
+ we will set new conditions on the mentioned workloads.
+- Downgrades
+ When downgrading from a release with this feature, to a release without, 
+ we will remove the extra conditions from workload objects.
+
+### Version Skew Strategy
+
+None identified yet.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: ExtendedWorkloadConditions
+  - Components depending on the feature gate:
+    - kube-controller-manager
+
+###### Does enabling the feature change any default behavior?
+No. The default behavior won't change. Only new conditions will be added with no effect on existing conditions.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+ Yes. Using the featuregates is the only way to enable/disable this feature
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The controllers start updating the new workload conditions again.
```

Comment:
> Is the API server currently preventing these new fields from being set on non-deployment applications? If so, it will also need a feature flag in order to change the behaviour.

No, currently the API server is not preventing these changes and will just copy the status update. The status is reconciled by each workload controller, thus a feature flag in API server shouldn't be necessary.